README.md
# W20 simple theme [![Build status](https://travis-ci.org/seedstack/w20-simple-theme.svg?branch=master)](https://travis-ci.org/seedstack/w20-business-theme)

Simple W20 theme with plain bootstrap styling and a basic top menu.

# Copyright and license

This source code is copyrighted by [The SeedStack Authors](https://github.com/seedstack/seedstack/blob/master/AUTHORS) and
released under the terms of the [Mozilla Public License 2.0](https://www.mozilla.org/MPL/2.0/).


CHANGELOG.md
# Version 3.3.1 (2018-07-18)

* [fix] Fix incompatibility with W20 2.4.2
* [fix] Fix views menu dropdown error 

# Version 3.3.0 (2018-03-21)

* [brk] Switched from Bower to NPM package manager
* [chg] Update W20 to ^2.4.0
* [chg] Update W20 Bootstrap 3 to ^2.3.0

# Version 3.2.2 (2016-10-06)

* [fix] Fix topbar links when application has a non-empty base href.

# Version 3.2.1 (2016-09-16)

* [fix] Fix menu links which were not honoring the route target attribute. 
* [fix] Fix topbar logo link which was reloading the whole application when in HTML5 mode.

# Version 3.2.0 (2016-04-27)

* [new] Add support for topbar logo.
* [chg] Replace the views by a more elegant multi-level dropdown.
* [chg] Update for SeedStack 16.4.
* [brk] The `menu` module has been renamed to `main`.

# Version 3.1.1 (2015-12-09)

* [chg] Update bower.json to accept latest minor versions of dependencies (^).

# Version 3.1.0 (2015-11-17)

* [brk] Refactored as an add-on.

# Version 3.0.0 (2015-07-28)

* [new] Initial release.


modules/main.js
/*
 * Copyright (c) 2013-2016, The SeedStack authors <http://seedstack.org>
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

define([
    'require',
    'module',

    '{lodash}/lodash',
    '{angular}/angular',

    '[text]!{w20-simple-theme}/templates/topbar.html',

    '{angular-sanitize}/angular-sanitize',
    '{w20-core}/modules/ui',
    '{w20-core}/modules/culture',
    '{w20-core}/modules/utils'

], function (require, module, _, angular, topbarTemplate) {
    'use strict';

    var _config = module && module.config() || {},
        w20SimpleThemeMain = angular.module('w20SimpleThemeMain', ['w20CoreCulture', 'w20CoreUtils', 'ngSanitize']),
        showTopbar = true,
        includes = _.contains || _.includes;

    w20SimpleThemeMain.directive('w20Topbar', ['$route', '$location', 'EventService', 'DisplayService', 'MenuService', 'EnvironmentService', 'ApplicationService', 'SecurityExpressionService', 'CultureService', 'NavigationService',
        function ($route, $location, eventService, displayService, menuService, environmentService, applicationService, securityExpressionService, cultureService, navigationService) {
            function isRouteVisible(route) {
                return !route.hidden && (typeof route.security === 'undefined' || securityExpressionService.evaluate(route.security));
            }

            return {
                template: topbarTemplate,
                replace: true,
                transclude: true,
                restrict: 'A',
                scope: true,
                link: function (scope, iElement, iAttrs) {
                    scope.hideViews = _config.hideViews || false;
                    scope.title = iAttrs.title || '';
                    scope.description = iAttrs.subtitle || '';
                    scope.navActions = menuService.getActions;
                    scope.navAction = menuService.getAction;
                    scope.displayName = cultureService.displayName;
                    scope.envtype = environmentService.environment;
                    scope.buildLink = navigationService.buildLink;
                    scope.logo = _config.logo;

                    scope.isTopbarDisplayed = function () {
                        return showTopbar;
                    };

                    scope.routeCategories = function () {
                        return _.sortBy(_.uniq(_.filter(_.map($route.routes, function (route) {
                            if (typeof route.category !== 'undefined' && route.category !== '__top' && isRouteVisible(route)) {
                                return route.category;
                            } else {
                                return null;
                            }
                        }), function (elt) {
                            return elt !== null && (typeof _config.categories !== 'undefined' ? includes(_config.categories, elt) : true);
                        })), function (elt) {
                            if (typeof _config.categories !== 'undefined') {
                                return _.indexOf(_config.categories, elt);
                            } else {
                                return elt;
                            }
                        });
                    };

                    scope.topLevelRoutes = function () {
                        return _.filter($route.routes, function (route) {
                            return route.category === '__top' && isRouteVisible(route);
                        });
                    };

                    scope.routesFromCategory = function (category) {
                        return _.filter($route.routes, function (route) {
                            return route.category === category && isRouteVisible(route);
                        });
                    };

                    scope.routeSortKey = function (route) {
                        return route.sortKey || route.path;
                    };

                    displayService.registerContentShiftCallback(function () {
                        return [showTopbar ? 50 : 0, 0, 0, 0];
                    });
                }
            };
        }]);

    w20SimpleThemeMain.run(['$rootScope', 'DisplayService', 'MenuService', function ($rootScope, displayService, menuService) {
        $rootScope.$on('$routeChangeSuccess', function (event, routeInfo) {
            if (routeInfo && routeInfo.$$route) {
                switch (routeInfo.$$route.navigation) {
                    case 'none':
                        showTopbar = false;
                        break;
                    case 'topbar':
                        showTopbar = true;
                        break;
                    case 'full':
                    /* falls through */
                    default:
                        showTopbar = true;
                        break;
                }

                displayService.computeContentShift();
            }
        });

        if (!_config.hideSecurity) {
            menuService.addAction('login', 'w20-login', {
                sortKey: 0
            });
            menuService.addAction('logout', 'w20-logout', {
                sortKey: 100
            });
        }

        if (!_config.hideCulture) {
            menuService.addAction('culture', 'w20-culture', {
                sortKey: 200
            });
        }

        if (!_config.hideConnectivity) {
            menuService.addAction('connectivity', 'w20-connectivity', {
                sortKey: 300
            });
        }

        _.each(_config.links, function (link, idx) {
            if (idx < 10) {
                menuService.addAction('link-' + idx, 'w20-link', _.extend(link, {
                    sortKey: 400 + idx
                }));
            }
        });
    }]);

    return {
        angularModules: ['w20SimpleThemeMain'],
        lifecycle: {
            pre: function (modules, fragments, callback) {
                angular.element('body').addClass('w20-top-shift-padding w20-right-shift-padding w20-bottom-shift-padding w20-left-shift-padding');
                callback(module);
            }
        }
    };
});

deploy_ghpages.sh
#!/bin/sh
#
# Copyright (c) 2013-2015 by The SeedStack authors. All rights reserved.
#
# This file is part of SeedStack, An enterprise-oriented full development stack.
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
#


(
	mkdir build
    cp -r bower_components build/
    cp -r w20-simple-theme-demo build/
    cp index.html build/
    cp w20.app.json build/
    cd build
    git init
	git config user.name "Travis-CI"
	git config user.email "travis@seedstack.org"
	git add .
	git add bower_components -f
	git commit -m "Built for gh-pages of http://seedstack.github.io/w20-simple-theme"
	git push --force --quiet "https://${GITHUB_TOKEN}@github.com/seedstack/w20-simple-theme" master:gh-pages >/dev/null 2>&1
)

modules/menu.js
/*
 * Copyright (c) 2013-2016, The SeedStack authors <http://seedstack.org>
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

define([
    'require',
    'module',

    '{lodash}/lodash',
    '{angular}/angular',

    '[text]!{w20-simple-theme}/templates/topbar.html',

    '{angular-sanitize}/angular-sanitize',
    '{w20-core}/modules/ui',
    '{w20-core}/modules/culture',
    '{w20-core}/modules/utils'

], function (require, module, _, angular, topbarTemplate) {
    'use strict';

    var _config = module && module.config() || {},
        w20SimpleThemeMenu = angular.module('w20SimpleThemeMenu', ['w20CoreCulture', 'w20CoreUtils', 'ngSanitize']),
        showTopbar = true;

    w20SimpleThemeMenu.directive('w20Topbar', ['$route', '$location', 'EventService', 'DisplayService', 'MenuService', 'EnvironmentService', 'ApplicationService', 'SecurityExpressionService', 'CultureService',
        function ($route, $location, eventService, displayService, menuService, environmentService, applicationService, securityExpressionService, cultureService) {
            function isRouteVisible(route) {
                return !route.hidden && (typeof route.security === 'undefined' || securityExpressionService.evaluate(route.security));
            }

            return {
                template: topbarTemplate,
                replace: true,
                transclude: true,
                restrict: 'A',
                scope: true,
                link: function (scope, iElement, iAttrs) {
                    scope.linkPrefix = $location.$$html5 ? '' : '#!';
                    scope.homePath = $location.$$absUrl;
                    scope.hideViews = _config.hideViews || false;
                    scope.title = iAttrs.title || '';
                    scope.description = iAttrs.subtitle || '';
                    scope.navActions = menuService.getActions;
                    scope.navAction = menuService.getAction;
                    scope.displayName = cultureService.displayName;
                    scope.envtype = environmentService.environment;
                    scope.logoUrl = _config.logoUrl;
                    scope.logoImg = _config.logoImg;

                    scope.isTopbarDisplayed = function () {
                        return showTopbar;
                    };

                    scope.routeCategories = function () {
                        return _.sortBy(_.uniq(_.filter(_.map($route.routes, function (route) {
                            if (typeof route.category !== 'undefined' && route.category !== '__top' && isRouteVisible(route)) {
                                return route.category;
                            } else {
                                return null;
                            }
                        }), function (elt) {
                            return elt !== null && (typeof _config.categories !== 'undefined' ? _.contains(_config.categories, elt) : true);
                        })), function (elt) {
                            if (typeof _config.categories !== 'undefined') {
                                return _.indexOf(_config.categories, elt);
                            } else {
                                return elt;
                            }
                        });
                    };

                    scope.topLevelRoutes = function () {
                        return _.filter($route.routes, function (route) {
                            return route.category === '__top' && isRouteVisible(route);
                        });
                    };

                    scope.routesFromCategory = function (category) {
                        return _.filter($route.routes, function (route) {
                            return route.category === category && isRouteVisible(route);
                        });
                    };

                    scope.routeSortKey = function (route) {
                        return route.sortKey || route.path;
                    };

                    displayService.registerContentShiftCallback(function () {
                        return [showTopbar ? 50 : 0, 0, 0, 0];
                    });
                }
            };
        }]);

    w20SimpleThemeMenu.run(['$rootScope', 'DisplayService', 'MenuService', function ($rootScope, displayService, menuService) {
        $rootScope.$on('$routeChangeSuccess', function (event, routeInfo) {
            if (routeInfo && routeInfo.$$route) {
                switch (routeInfo.$$route.navigation) {
                    case 'none':
                        showTopbar = false;
                        break;
                    case 'topbar':
                        showTopbar = true;
                        break;
                    case 'full':
                    /* falls through */
                    default:
                        showTopbar = true;
                        break;
                }

                displayService.computeContentShift();
            }
        });

        if (!_config.hideSecurity) {
            menuService.addAction('login', 'w20-login', {
                sortKey: 0
            });
            menuService.addAction('logout', 'w20-logout', {
                sortKey: 100
            });
        }

        if (!_config.hideCulture) {
            menuService.addAction('culture', 'w20-culture', {
                sortKey: 200
            });
        }

        if (!_config.hideConnectivity) {
            menuService.addAction('connectivity', 'w20-connectivity', {
                sortKey: 300
            });
        }

        _.each(_config.links, function (link, idx) {
            if (idx < 10) {
                menuService.addAction('link-' + idx, 'w20-link', _.extend(link, {
                    sortKey: 400 + idx
                }));
            }
        });
    }]);

    return {
        angularModules: ['w20SimpleThemeMenu'],
        lifecycle: {
            pre: function (modules, fragments, callback) {
                angular.element('body').addClass('w20-top-shift-padding w20-right-shift-padding w20-bottom-shift-padding w20-left-shift-padding');
                callback(module);
            }
        }
    };
});